core/sql/optimizer/IndexDesc.cpp

```diff
@@ -802,10 +802,30 @@ IndexProperty::compareIndexPromise(const IndexProperty *ixProp) const
 
     const IndexDesc * index = getIndexDesc();
     const IndexDesc * otherIndex = ixProp->getIndexDesc();
-    if ( ((IndexColumn *)(index->getIndexKey()[0]).getItemExpr())->getDefinition() != 
-         ((IndexColumn *)(otherIndex->getIndexKey()[0]).getItemExpr())->getDefinition() )
-
-      return INCOMPATIBLE;
+    
+    // If the two indexes have differing leading columns, consider them incompatible.
+    // For this check, we ignore the "_SALT_" column if both are salted.
+    CollIndex columnToCheck = 0;
+    NABoolean done = FALSE;
+    while (!done)
+      {
+        if (columnToCheck >= index->getIndexKey().entries())
+          return INCOMPATIBLE;  // must be one of the indexes is just "_SALT_" (seems unlikely actually)
+        else if (columnToCheck >= otherIndex->getIndexKey().entries())
+          return INCOMPATIBLE;  // must be one of the indexes is just "_SALT_" (seems unlikely actually)
+        else
+          {
+            IndexColumn * indexCol = (IndexColumn *)(index->getIndexKey()[columnToCheck]).getItemExpr();
+            IndexColumn * otherIndexCol = (IndexColumn *)(otherIndex->getIndexKey()[columnToCheck]).getItemExpr();
+            if ( indexCol->getNAColumn()->isSaltColumn() &&
+                 otherIndexCol->getNAColumn()->isSaltColumn() )
```

Comment:
Looks good. I am wondering whether we should do the same thing with divisioning columns (method NAColumn::isDivisioningColumn()).

Reply:
I considered that. I think the intent of the code in IndexProperty::compareIndexPromise is, for each user column, to retain the "best" (in some sense) index that begins with that column. The salt column is not a user column in any sense; with the exception of pre-partitioning rows across regions, a user would never have a predicate on salt. A divisioning column is a fuzzy area. It is a system-generated column, but it reflects true user content. A user predicate may well result in a predicate on the divisioning column. So, I thought it was more in tune with the spirit of IndexProperty::compareIndexPromise to treat divisioning columns like user columns. But I do not feel strongly about it; a case could be made the other way.

Reply:
Just wanted to bring this up, I can also live with both solutions. The computation of predicates for salt and division columns is similar, both are computed from user columns. The main difference is that the expressions of divisioning columns are monotonically increasing. I would have thought that the following indexes should be treated as incompatible:

create index ix1 on t(a) division like table;
create index ix2 on t(b) division like table;

Reply:
It depends. Suppose the DIVISION BY clause is on C. Should we think of ix1 as an index on C, A or just on A? If I have a query, select c from table where c = value, then ix1 and ix2 are compatible for that query. Today, IndexProperty::compareIndexPromise doesn't take the select predicates into account; it is just looking for the "best" index on C. (And on A. And on B.)

Reply:
IMHO the user should consider this a partitioned/divisioned index on A and not an index on (C, A), and so should we.

Reply:
OK. Now a question: Can an index have a different divisioning column than the base table? Or can I assume that if NAColumn::isDivisioningColumn() returns true for both index and otherIndex that they are the same column?

Reply:
Right now, all we allow is DIVISION LIKE TABLE, the index must have the same divisioning columns as the base table, if any. So, it's the same restriction we have for salting.